### DISCUSSION TITLE
Multihead replay finetuning converges more slowly than regular training

### DISCUSSION THREAD

**Discussion by lucasdekam** [ACEsuit/mace#626] [Q&A]
Hi, I'd like to share my experience so far with multihead finetuning and ask for ideas.

I'm training MACE on a dataset with ~80 platinum-water interface structures of ~400 atoms each; energy and forces are evaluated using VASP with the RPBE functional. I've tried two methods: multihead replay finetuning (starting from the mace-mp0b agnesi small model) and naive finetuning (starting from the standard small model). The training parameters can be found below. The multihead finetuning converges much more slowly and the model struggles to converge the errors on the replayed data (pt_head) again. I've not been able to get the forces error much lower than 100 meV/A, although perhaps this could be achieved by longer training. 

```
2024-10-06 18:06:17.060 INFO: Epoch 99: head: pt_head, loss=  0.0016, RMSE_E_per_atom=   639.6 meV, RMSE_F=   729.4 meV / A, RMSE_stress=    40.4 meV / A^3
2024-10-06 18:06:17.189 INFO: Epoch 99: head: default, loss=  0.0026, RMSE_E_per_atom=     1.2 meV, RMSE_F=   119.0 meV / A, RMSE_stress=    10.9 meV / A^3
``` 

Because I wanted to converge the forces faster, I increased the forces weight by a factor 10, which gave this result. With equal weights the forces converge even more slowly.

On the other hand, naive finetuning converges pretty fast to a a rather low force RMSE. For me, the resulting model also seems very stable, so there's no "catastrophic forgetting" (at least not that I've noticed). 

```
2024-10-06 18:41:18.014 INFO: Epoch 49: head: default, loss=  0.0009, RMSE_E_per_atom=     0.3 meV, RMSE_F=    51.1 meV / A
```

I'm still interested in using the multiheads training, as it might improve the generalizability of my model. My question: what could cause the slow convergence of multiheads training? Is this already known? What parameters can one tune to achieve better convergence (should I increase the forces weight even more, etc.)? I'd be happy to hear about any insights :)

### Training parameters:

Multihead
```
mace_run_train \
    --name="MACE" \
    --foundation_model="../mace_agnesi_small.model" \
    --multiheads_finetuning=True \
    --train_file="../train.xyz" \
    --valid_fraction=0.05 \
    --test_file="../test.xyz" \
    --energy_weight=1.0 \
    --forces_weight=10.0 \
    --energy_key='DFT_energy' \
    --forces_key='DFT_forces' \
    --E0s="{1: -1.20502718, 8: -1.60386686, 78: -0.5578757}" \
    --lr=0.01 \
    --scaling="rms_forces_scaling" \
    --batch_size=3 \
    --max_num_epochs=100 \
    --ema \
    --ema_decay=0.99 \
    --amsgrad \
    --default_dtype="float64" \
    --device=cuda \
    --seed=3
```

Naive
```
mace_run_train \
    --name="MACE" \
    --foundation_model="small" \
    --multiheads_finetuning=False \
    --train_file="../train.xyz" \
    --valid_fraction=0.05 \
    --test_file="../test.xyz" \
    --energy_weight=1.0 \
    --forces_weight=1.0 \
    --energy_key='DFT_energy' \
    --forces_key='DFT_forces' \
    --E0s="{1: -1.20502718, 8: -1.60386686, 78: -0.5578757}" \
    --lr=0.01 \
    --scaling="rms_forces_scaling" \
    --batch_size=2 \
    --max_num_epochs=50 \
    --ema \
    --ema_decay=0.99 \
    --amsgrad \
    --default_dtype="float64" \
    --device=cuda \
    --seed=1
```

**Comment by ilyes319** [comment]
Hello,

Can you please share the log files for the two training so I can help you. I need for example to look at the initial loss to see if there is a potential problem.
The multihead replay requires precise computation of the E0s, and that is usually the reason for problems.
How did you compute the E0s for your DFT. Did you make sure that the oxygen E0s are spin polarized, it is very important.

**Reply by lucasdekam** [reply]
Hi Ilyes, I've added the two log files below.

I computed the E0s with a spin polarized computation. The RPBE energies are pretty close to the PBE E0s from the MP dataset (see multihead.log file). This is the VASP INCAR for an E0 computation:

```
INCAR created by Atomic Simulation Environment
 ENCUT = 400.000000
 SIGMA = 0.050000
 EDIFF = 1.00e-06
 GGA = RP
 ISMEAR = -1
 ISPIN = 2
 NCORE = 8
 LCHARG = .FALSE.
 LWAVE = .FALSE.
```

[multihead.log](https://github.com/user-attachments/files/17325388/multihead.log)
[naive.log](https://github.com/user-attachments/files/17325391/naive.log)

**Reply by ilyes319** [reply]
Indeed the E0s look fine. 
I have two hypothesis atm:
1. Could you try to increase the batchsize a bit? At least going to 3. I know you have very large configs but that would help with having diverse enough steps.
2. Could you try putting the stress weight to zero for the multihead finetuning? With the naive finetuning, you are not training on stress, and I am wondering if this could effect the training.

**Reply by lucasdekam** [reply]
Thanks for your tips. I've played around with both the batchsize and the stress weight now. Increasing the batch size to 3 or 4 seems to have no significant effect.

Regarding the stress weight I now added `--stress_weight=0` but the training does keep printing the RMSE_stress each epoch. Is that expected behavior? How does it compute the stress RMSE on my configs, which does not include any stress data? 

This is the extxyz format of my training configs, in case that helps for context: 
```
Lattice="11.144002871499989 0.0 0.0 0.0 9.650989586565721 0.0 0.0 0.0 41.37380030303563" Properties=species:S:1:pos:R:3:DFT_forces:R:3 structure=interface type=md_400K id=60 DFT_energy=-2032.27413192 pbc="T T T"
```

**Reply by ilyes319** [reply]
Hey, indeed if you don't have stress data, it should be ignoring the stress anyway. It will print it atm because of how the finetuning is handled but it should be ignoring it. For extra safety, do put `--stress_weight=0`.

**Reply by lucasdekam** [reply]
In your experience, does the RMSE of both heads become satisfactory eventually?

I also noticed `2024-10-08 12:13:29.370 INFO: Number of configurations after filtering 85 is less than the number of samples 1000, selecting random configurations for the rest.` in the log, does this mean there's not so many configs with Pt, H and O in the MP0 training set?

**Reply by ilyes319** [reply]
I expect the MP head to go up quite a lot but not to that extent.

Yes that means that not so many configs have combinations of Pt, H and O.

**Reply by lucasdekam** [reply]
Hmm, maybe it's difficult for the finetuning if the finetuning set & replay set have different elements?

**Reply by ilyes319** [reply]
It should be fine, we have quite a lot of experience with that. Could you try running the same run but with 3 different seeds and share the log files?

**Reply by lucasdekam** [reply]
Can do, any preference for the forces weight? Should I leave it at 10 or revert to 1?

**Reply by ilyes319** [reply]
I think use 10 forces and 1 energies, that is matching what we use for MP.

**Reply by lucasdekam** [reply]
Here's the result, with this training command:
```
mace_run_train \
    --name="MACE" \
    --foundation_model="../mace_agnesi_small.model" \
    --multiheads_finetuning=True \
    --train_file="../train.xyz" \
    --valid_fraction=0.1 \
    --test_file="../test.xyz" \
    --energy_weight=1.0 \
    --forces_weight=10.0 \
    --stress_weight=0 \
    --virials_weight=0 \
    --energy_key='DFT_energy' \
    --forces_key='DFT_forces' \
    --stress_key=None \
    --virials_key=None \
    --compute_stress=False \
    --E0s="{1: -1.20502718, 8: -1.60386686, 78: -0.5578757}" \
    --lr=0.01 \
    --scaling="rms_forces_scaling" \
    --batch_size=3 \
    --max_num_epochs=50 \
    --ema \
    --ema_decay=0.99 \
    --amsgrad \
    --default_dtype="float64" \
    --device=cuda \
    --seed=3
```
where only the seed was changed.
[MACE_run-2.log](https://github.com/user-attachments/files/17375333/MACE_run-2.log)
[MACE_run-3.log](https://github.com/user-attachments/files/17375334/MACE_run-3.log)
[MACE_run-1.log](https://github.com/user-attachments/files/17375335/MACE_run-1.log)

**Reply by ilyes319** [reply]
Hey @lucasdekam, thank you for the log. We do see a large variation with the seed. I will have a more proper look at it. To be more fair, it would be better if you create a single validation file, for all the seeds. Here, the variation could come the pretraining head but also the data in the validation set.

**Reply by lucasdekam** [reply]
[multiheads-0_run-0.log](https://github.com/user-attachments/files/17392234/multiheads-0_run-0.log)
[multiheads-1_run-1.log](https://github.com/user-attachments/files/17392235/multiheads-1_run-1.log)
[multiheads-2_run-2.log](https://github.com/user-attachments/files/17392236/multiheads-2_run-2.log)

I've done a quick run with a fixed validation set file. Hope that helps. Let me know if you find anything. 

```
NAME="multiheads-$SLURM_ARRAY_TASK_ID"

mace_run_train \
    --name=$NAME \
    --foundation_model="mace_agnesi_small.model" \
    --multiheads_finetuning=True \
    --train_file="train_multiheads.xyz" \
    --valid_file="valid_multiheads.xyz" \
    --test_file="test.xyz" \
    --energy_weight=1.0 \
    --forces_weight=10.0 \
    --stress_weight=0 \
    --energy_key='DFT_energy' \
    --forces_key='DFT_forces' \
    --E0s="{1: -1.20502718, 8: -1.60386686, 78: -0.5578757}" \
    --lr=0.01 \
    --scaling="rms_forces_scaling" \
    --batch_size=3 \
    --max_num_epochs=50 \
    --ema \
    --ema_decay=0.99 \
    --amsgrad \
    --default_dtype="float64" \
    --device=cuda \
    --seed=$SLURM_ARRAY_TASK_ID \
    --work_dir=$NAME
```

**Reply by Ssbban** [reply]
Hi,
Have you solved this problem? I have recently used the latest version of MACE0.3.12 and found that the multihead training speed is very slow and it also takes a longer time to converge energy and force.

**Reply by lucasdekam** [reply]
Hey, I haven't! I decided that training from scratch gives a good enough result for me. Thanks for posting your experience in the other thread (https://github.com/ACEsuit/mace/discussions/875 for future readers), I'm sure that helps the developers. Your atomic structures seem a lot smaller than mine, so it was not the big configs that caused my problem, it seems.
